### PR TITLE
Implement block parameter support for eth_estimateGas

### LINKED ISSUE #7414: `eth_estimateGas` does not respect the `block` parameter
### Description
As defined by the standard https://ethereum.github.io/execution-apis/api-documentation/ `eth_estimateGas` have a second optional parameter that specified the block to use for the estimation, we ignore that option and just use the chain head in every case.

### Acceptance Criteria

`eth_estimateGas` adhere to the standard and accepts and uses the value passed in the `block` parameter to perform the estimation.

### PR DESCRIPTION
**Pull request by 7suyash7**
Fixes #7414 

Implement support for the optional `block` parameter in the `eth_estimateGas` RPC method.

- Add a `blockParameter` method to parse the optional block parameter from the request.
- Modify `response` method to use the specified block for gas estimation.
- Implement `getBlockNumberFromParameter` method to convert block parameters to block numbers.
- Update existing tests to include the new block parameter functionality.
- Add new tests to verify behaviour with different block parameters.
- Update mocking in test methods to handle block number queries.

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java [hunk 1]
```diff
@@ -120,18 +137,15 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
...
-            .build(),
-        operationTracer,
-        blockHeader.getNumber());
+  private long getBlockNumberFromParameter(final BlockParameter blockParameter) {
+    if (blockParameter.isLatest()) {
+      return blockchainQueries.headBlockNumber();
+    } else if (blockParameter.isEarliest()) {
+      return BlockHeader.GENESIS_BLOCK_NUMBER;
+    } else if (blockParameter.isPending()) {
+      return blockchainQueries.headBlockNumber() + 1;
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java:146]: is this what other methods do for pending? I thought pending had the same behavior as latest elsewhere

**Review by macfarla — COMMENTED** [review]
needs a changelog entry. also could add some http spec tests - see ethereum/api/src/test/resources/org/hyperledger/besu/ethereum/api/jsonrpc/eth/eth_estimateGas_noParams.json

**Comment by fab-10** [comment]
Thanks for working on this! 
Have you considered extending from `AbstractBlockParameterMethod`?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java [hunk 2]
```diff
@@ -74,26 +81,33 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
...
-        executeSimulation(
-            blockHeader,
+        transactionSimulator.process(
             overrideGasLimitAndPrice(callParams, low),
+            ImmutableTransactionValidationParams.builder()
+                .from(TransactionValidationParams.transactionSimulator())
+                .isAllowExceedingBalance(isAllowExceedingBalance)
+                .build(),
             operationTracer,
-            isAllowExceedingBalance);
```
[the remark this answers is not in the document]
Comment (7suyash7) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java:96]: I haven't actually removed the `isAllowExceedingBalance` logic, I just removed the `executeSimulation` method and now call `transactionSimulator.process` directly. `isAllowExceedingBalance` is passed as a parameter of the `TransactionValidationParams` to `transactionSimulator.process`.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java (hunk 1 above)
Comment (7suyash7) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java:146]: Yeah, I just noticed that, will change it.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java [hunk 3]
```diff
@@ -29,18 +30,13 @@
...
-import java.util.Optional;
-
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class EthEstimateGas extends AbstractEstimateGas {
 
-  private static final Logger LOG = LoggerFactory.getLogger(EthEstimateGas.class);
-
   public EthEstimateGas(
```
[the remark this answers is not in the document]
Comment (7suyash7) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java:40]: Yeah, I did consider it, but `AbstractEstimateGas` provides functionality specific to gas estimation. In the current implementation block parameter handling logic mirrors the approach used in `AbstractBlockParameterMethod`. If using `AbstractBlockParameterMethod` is a standard, I can refactor it no worries.

**Comment by fab-10** [comment]
Closing this as replaced by https://github.com/hyperledger/besu/pull/7502

**Pull request closed without merging**